here-naksha-lib-handlers/src/main/java/com/here/naksha/lib/handlers/SourceIdHandler.java

```diff
@@ -60,70 +60,77 @@ public SourceIdHandler(
     final Request<?> request = event.getRequest();
     logger.info("Handler received request {}", request.getClass().getSimpleName());
     if (request instanceof ReadFeatures readRequest) {
-      Optional.ofNullable(readRequest.getPropertyOp())
-          .ifPresent(t -> PropertyOperationUtil.replacePropertyInPropertyOperationTree(
-              t, SourceIdHandler::transformPopWithSourceId));
+      transformPropertyOperation(readRequest);
     } else if (request instanceof WriteXyzFeatures writeRequest) {
       writeRequest.features.stream()
           .map(XyzFeatureCodec::getFeature)
           .filter(Objects::nonNull)
-          .peek(this::removeSourceIdTagsIfAny)
           .forEachOrdered(this::setSourceIdTags);
     }
 
     return event.sendUpstream(request);
   }
 
+  private void transformPropertyOperation(ReadFeatures readRequest) {
+
+    if (readRequest.getPropertyOp() == null) {
+      return;
+    }
+
+    POp propertyOp = readRequest.getPropertyOp();
+
+    if (propertyOp.children() != null) {
```

Comment:
Would it make sense to also verify whether `children` is `notEmpty`?

Reply:
Yes sure.